### PR TITLE
Support equi joins in MV Plan Validator

### PR DESCRIPTION
**Pull request by gracexin2003**
This branch supports equi joins in the MaterializedViewPlanValidator class.
The functions help support join commands with the USING clause, and
check for nested equi join conditions.

Resolves: #16033 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java [hunk 1]
```diff
@@ -74,9 +89,29 @@ protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, Materi
         if (!node.getOperator().equals(LogicalBinaryExpression.Operator.AND)) {
             throw new SemanticException(NOT_SUPPORTED, node, "Only AND operator is supported for join criteria for materialized view.");
         }
+
+        visitLogicalBinaryExpressionChild(node.getLeft(), context);
+        visitLogicalBinaryExpressionChild(node.getRight(), context);
+
```
Comment (lliiun-z) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:95]: Good, this solves the recursive problem. Please remember we still need to check if "=" exists in the leaf node.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java [hunk 2]
```diff
@@ -53,12 +59,21 @@ protected Void visitJoin(Join node, MaterializedViewPlanValidatorContext context
         }
 
         JoinCriteria joinCriteria = node.getCriteria().get();
-        if (!(joinCriteria instanceof JoinOn)) {
-            throw new SemanticException(NOT_SUPPORTED, node, "Only join-on is supported for materialized view.");
+        if (!(joinCriteria instanceof JoinOn) && !(joinCriteria instanceof JoinUsing)) {
```
Comment (lliiun-z) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:62]: Great!

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java [hunk 3]
```diff
@@ -74,9 +89,29 @@ protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, Materi
...
+        if (node instanceof ComparisonExpression) {
+            if (node instanceof LogicalBinaryExpression) {
+                visitLogicalBinaryExpression((LogicalBinaryExpression) node, context);
+            }
+            else {
+                throw new SemanticException(NOT_SUPPORTED, node, "Only comparison or logical binary expressions are supported for join criteria for materialized view.");
+            }
+        }
+    }
+
```
Comment (gggrace14) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:114]: Pls check on the body of ComparisonExpression.

A better way is to implement the visitComparisonExpression() method, ex.,

```suggestion
    @Override
    protected Void visitComparisonExpression(ComparisonExpression node, MaterializedViewPlanValidatorContext context)
    {
        if (!context.isProcessingJoinNode()) {
            return super.visitComparisonExpression(node, context);
        }

        if (!node.getOperator().equals(ComparisonExpression.Operator.EQUAL)) {
            throw new SemanticException(NOT_SUPPORTED, node, "Only EQUAL join is supported for materialized view.");
        }
        return super.visitComparisonExpression(node, context);
   }

```

**Review by gggrace14 — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java (hunk 1 above)
Comment (gggrace14) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:95]: Yes, Lin is correct that we do want to check the leaf node is "=" ComparisonExpression but nothing else. The standard way should be to implement the ComparisonExpression interface, and check the Operator.EQUAL there.

Separately, if you look into the base class default implementation of super.visitLogicalBinaryExpression(node, context), you'll find you don't need the above two lines of visitLogicalBinaryExpressionChild. Calling super.visitLogicalBinaryExpression(node, context) is sufficient. Sorry, we weren't very clear in the Issue description.

> 

@Override
protected R visitLogicalBinaryExpression(LogicalBinaryExpression node, C context)
{
    process(node.getLeft(), context);
    process(node.getRight(), context);

    return null;
}
If the node.getLeft() is a LogicalBinaryExpression, process(node.getLeft()) will call visitLogicalBinaryExpression() on node.getLeft() recursively.

If you follow more on process(Node node), you will see it eventually calls visitT() where T is the actual type/class of node. This idea, as well as the recursion idea Lin mentioned, is very common for AST visitor implementation, in general.

**Comment by gggrace14** [comment]
Also @gracexin2003 , I notice you have 2 commits in the PR and one of them is from merge. That is not correct. It should only include your commit. You should fix your local git branches. In your local, the branches should lay out similar to https://pxl.cl/1KgHG

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java [hunk 4]
```diff
@@ -74,9 +78,24 @@ protected Void visitLogicalBinaryExpression(LogicalBinaryExpression node, Materi
         if (!node.getOperator().equals(LogicalBinaryExpression.Operator.AND)) {
             throw new SemanticException(NOT_SUPPORTED, node, "Only AND operator is supported for join criteria for materialized view.");
         }
+
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:81]: remove this line

**Review by highker — COMMENTED** [review]
@gggrace14, could you review this PR and let me know once it's ready?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java (hunk 4 above)
Comment (gggrace14) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java:81]: Hi @gracexin2003 , it looks great to me and approving. Can you help address James' comment here and remove this redundant empty line?

**Review by gggrace14 — APPROVED** [review]

**Comment by gggrace14** [comment]
> @gggrace14, could you review this PR and let me know once it's ready?

Got it!

**Comment by gggrace14** [comment]
Hi @gracexin2003 , need to get all checks pass in order to have the "Merge" button green. The test failure might not be related to your change. Just try to change the commit message and keep pushing, until all checks pass :)

**Review by highker — APPROVED** [review]

**Comment by lliiun-z** [comment]
Thanks @gggrace14 @highker for the review!

**Comment by lliiun-z** [comment]
@gggrace14 , do you think if the Facebook Integration check failure matters?

**Comment by highker** [comment]
Let's rebase and push. The FB failure could be due to @gggrace14's recent commits

**Comment by gggrace14** [comment]
Yeah, it seems like we need to rebase and push again, for the Facebook Integration to pass. We might not be able to "merge" if one check doesn't pass. @gracexin2003 @lliiun-z 

I retriggered Jenkins, and got the same error

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project presto-prism: Compilation failure: Compilation failure: 
[ERROR] /tmp/workspace/presto-facebook-oss-integration-github/presto-prism/src/main/java/com/facebook/presto/hive/PrismMetadata.java:[1003,4] error: method does not override or implement a method from a supertype
[ERROR] /tmp/workspace/presto-facebook-oss-integration-github/presto-prism/src/main/java/com/facebook/presto/hive/PrismMetadata.java:[1008,71] error: cannot find symbol
[ERROR]   symbol:   method beginRefreshMaterializedView(ConnectorSession,ConnectorTableHandle)
[ERROR]   location: variable hiveMetadata of type HiveMetadata

**Comment by highker** [comment]
The PR is still not rebased. Can we pull the master, rebase, and push?

**Review by jainxrohit — APPROVED** [review]
Joining the party late, this is already in great shape. Thanks for working on it!!

**Comment by highker** [comment]
@gracexin2003, sorry, one last nit I missed: for the commit message title "Support equi joins in MV Plan Validator", we need to be explicit without abbrvs: "Support equi joins in materialized view validator"

**Pull request merged**